web_external/util/annotationGeometryParser.js

```diff
@@ -1,154 +1,261 @@
+//=============================================================================
+class AnnotationGeometryTrack {
+    //-------------------------------------------------------------------------
+    constructor() {
+        this._frameRange = [0, 0];
+        this.enableState = true;
+    }
+
+    //-------------------------------------------------------------------------
+    expandFrameRange(frame) {
+        this._frameRange[0] = Math.min(this._frameRange[0], frame);
+        this._frameRange[1] = Math.max(this._frameRange[1], frame);
+    }
+
+    //-------------------------------------------------------------------------
+    get frameRange() {
+        return this._frameRange;
+    }
+}
+
+//=============================================================================
 class AnnotationGeometryContainer {
+    //-------------------------------------------------------------------------
     constructor() {
         this._itemId = null;
         this._id0 = 0;
-        this._frameMap = new Map();
-        this._editedGeom = new Set();
-        this._addedGeom = new Set();
-        this._trackIds = new Set();
-        this._trackRanges = new Map();
-        this._enableState = new Map();
+
+        this._geomMap = new Map(); // {track id, frame} -> id0
+        this._trackMap = new Map(); // track id -> AnnotationGeometryTrack
+        this._frameMap = new Map(); // frame -> Map(track id -> geometries)
+
+        this._addedGeom = new Map();
+        this._editedGeom = new Map();
+        this._removedGeom = new Map();
     }
 
+    //-------------------------------------------------------------------------
     add(geometry) {
         this._itemId = geometry.itemId;
         this._id0 = Math.max(this._id0, geometry.id0);
-        var frameMap = this._frameMap;
-        if (!frameMap.has(geometry.ts0)) {
-            frameMap.set(geometry.ts0, []);
-        }
-        frameMap.get(geometry.ts0).push(geometry);
-        if (!this._trackIds.has(geometry.id1)) {
-            this._trackIds.add(geometry.id1);
-            this._enableState.set(geometry.id1, true);
+
+        // Create frame if needed
+        if (!this._frameMap.has(geometry.ts0)) {
+            this._frameMap.set(geometry.ts0, new Map());
```

Comment:
I think the map of map is harder to understand than it was before with a flat organization.

Reply:
...but it's *much* easier to *use*. It avoids several instances that before would have had to scan the whole array looking for a specific state. Now, I can just use key-lookup to directly access the desired value.

For example:
```javascript
    var geomToChange = Array.from(geomSet).find((geom) => {
        return geom.id1 === trackId && geom.ts0 === frame;
    });
```

...becomes:
```javascript
    var geomToChange = geomSet.get(trackId);
```

(...except that particular code wound up using the state map instead, but you get the idea.)

It may be (slightly) harder to understand up front, but it makes it *easier* to understand the code in many places the map is actually used.